src/exchanges/bitmex/BitmexMarketFeed.ts

```diff
@@ -0,0 +1,190 @@
+        }
+        super(config);
+        this.owner = 'BitMEX';
+        this.feedUrl = config.wsUrl;
+        this.seq = 0;
+        this.connect();
+    }
+
+    public subscribe(productIds: string[]) {
+        this.logger.log('debug', `Subscribing to the following symbols: ${JSON.stringify(productIds)}`);
+        productIds.forEach((productId: string) => {
+            const subscribeMessage = {
+                op: 'subscribe',
+                args: [`orderBookL2:${productId}`, `trade:${productId}`],
+            };
+
+            this.send(JSON.stringify(subscribeMessage));
+        });
+    }
+
+    protected onOpen(): void {
+        // Nothing for now
+    }
+
+    protected handleMessage(rawMsg: string): void {
+        const msg: BitmexMessage = JSON.parse(rawMsg) as BitmexMessage;
+
+        if (msg.error) {
+            const err = new Error(`Error while subscribing to symbols: ${msg.error}`);
+            this.onError(err);
+        } else if (msg.table === 'trade') {
+            // trade message
+            const tradeMsg: BitmexTradeMessage = msg as BitmexTradeMessage;
+            this.handleTrade(tradeMsg);
+        } else if (msg.action) {
+            if (msg.action === 'partial') {
+                // orderbook snapshot
+                const snapshotMsg = msg as OrderbookSnapshotMessage;
+                this.handleSnapshot(snapshotMsg);
+            } else {
+                // orderbook update
+                const updateMsg = msg as OrderbookUpdateMessage;
+                this.handleOrderbookUpdate(updateMsg);
+            }
+        } else if (msg.success !== undefined) {
+            // subscription response
+            const subscriptionResMSg: SubscriptionResponseMessage = msg as SubscriptionResponseMessage;
+            this.handleSubscriptionSuccess(subscriptionResMSg);
+        } else if (msg.info) {
+            // welcome message
+            this.logger.log('debug', 'Received welcome message from BitMEX WS feed.');
+        } else {
+            // unhandled/unexpected message
+            const err = new Error(`An unhandled message was received over the websocket connection: ${JSON.stringify(rawMsg)}`);
+            this.onError(err);
+        }
+    }
+
+    /**
+     * Gets the next sequence number, incrementing it for the next time it's called.
+     */
+    private getSeq(): number {
+        this.seq += 1;
+        return this.seq;
+    }
+
+    private handleSnapshot(snapshot: OrderbookSnapshotMessage) {
+        // (re)initialize our order id map
+        const newIdMap = snapshot.data.reduce((acc, { id, price }) => ({...acc, [id]: price }), {});
+        this.orderIdMap = newIdMap;
+
+        const mapLevelUpdates: (date: PriceData) => PriceLevelWithOrders =
+            ({ id, price, size, side }) => PriceLevelFactory(price, size, side.toLowerCase());
+
+        const asks: PriceLevelWithOrders[] = snapshot.data
+            .filter(R.propEq('side', 'Sell'))
```

Comment:
We try not to add libraries unnecessarily. This is the only place you use ramda and can easily be replaced by the native `Array.filter` method.